### PR TITLE
move tick generation back to banking_stage, add unit tests

### PR DESCRIPTION
**Pull request by rob-solana**

### CODE REVIEW

File: src/banking_stage.rs [hunk 1]
```diff
@@ -241,63 +239,163 @@ impl Service for BankingStage {
...
+
+        // glad they all fit
+        assert_eq!(packets.len(), 1);
+        verified_sender                       // tx, no_ver, anf
+            .send(vec![(packets[0].clone(), vec![1u8, 0u8, 1u8])])
+            .unwrap();
+
+        drop(verified_sender);
+
+        let entries: Vec<_> = entry_receiver.iter().map(|x| x).collect();
```
Comment (garious) [inline src/banking_stage.rs]: Looks like you should be able to change this to `flat_map()` and remove the loop below it.

File: src/banking_stage.rs [hunk 2]
```diff
@@ -241,63 +239,163 @@ impl Service for BankingStage {
...
+        // Assert the user holds one token, not two. If the stage only outputs one
+        // entry, then the second transaction will be rejected, because it drives
+        // the account balance below zero before the credit is added.
+        let bank = Bank::new(&mint);
+        for entries in ventries {
+            for entry in entries {
+                assert!(
+                    bank.process_transactions(&entry.transactions)
+                        .into_iter()
+                        .all(|x| x.is_ok())
```
Comment (garious) [inline src/banking_stage.rs]: I see a few things in this test that I'd guess clippy would complain about. Have you looked at it? Or perhaps it doesn't reach our tests?

File: src/banking_stage.rs [hunk 3]
```diff
@@ -241,63 +239,163 @@ impl Service for BankingStage {
...
+    use mint::Mint;
+    use packet::{to_packets, PacketRecycler};
+    use signature::{Keypair, KeypairUtil};
+    use std::thread::sleep;
+    use transaction::Transaction;
+
+    #[test]
+    fn test_banking_stage_shutdown() {
+        use logger;
+        logger::setup();
```
Comment (garious) [inline src/banking_stage.rs]: Were you making use of the logger in any of these tests?

File: src/banking_stage.rs [hunk 4]
```diff
@@ -43,26 +42,39 @@ impl BankingStage {
...
-                        &hash_sender,
-                        &poh_receiver,
+                        &poh_service,
                         &verified_receiver,
                         &entry_sender,
                     ) {
                         match e {
                             Error::RecvTimeoutError(RecvTimeoutError::Disconnected) => break,
                             Error::RecvTimeoutError(RecvTimeoutError::Timeout) => (),
+                            Error::RecvError(_) => break,
```
Comment (garious) [inline src/banking_stage.rs]: Is this an artifact of the previous strategy or still relevant?

File: src/banking_stage.rs [hunk 5]
```diff
@@ -80,26 +92,33 @@ impl BankingStage {
             }).collect()
     }
 
+    fn tick(poh_service: &PohService, entry_sender: &Sender<Vec<Entry>>) -> Result<()> {
```
Comment (garious) [inline src/banking_stage.rs]: Can you use EntrySender here? https://github.com/solana-labs/solana/blob/master/src/entry.rs#L16

File: src/banking_stage.rs (hunk 4 above)
Comment (rob-solana) [reply src/banking_stage.rs]: I have a blocking recv() in process_packets()

File: src/banking_stage.rs (hunk 5 above)
Comment (rob-solana) [reply src/banking_stage.rs]: sure, didn't know about the type

it doesn't conflict with the "one file one style" rule?

File: src/banking_stage.rs (hunk 3 above)
Comment (rob-solana) [reply src/banking_stage.rs]: did use it, took it out once comfy with paths, will remove

File: src/banking_stage.rs (hunk 2 above)
Comment (rob-solana) [reply src/banking_stage.rs]: haven't looked at clippy, will do

File: src/banking_stage.rs (hunk 1 above)
Comment (rob-solana) [reply src/banking_stage.rs]: cool, will do

**Comment by aeyakovenko** [comment]
@rob-solana i dont fully grok why this needs to be synchronous.  bank did some work, hash gets encoded into poh, new last id is registered.  there is no way for anything to reference that new last id until its registered.

File: src/banking_stage.rs (hunk 2 above)
Comment (rob-solana) [reply src/banking_stage.rs]: clippy is happy with all these

**Comment by rob-solana** [comment]
@aeyakovenko the issue is that transactions that have been verified (but not yet recorded) against a last_id may be in flight, in channel queues, while transactions ahead of them being recorded push that last_id out of the list.

**Comment by garious** [comment]
@rob-solana, can we get this one merged?

**Pull request merged**